course5/homework10/linked_list.h

```diff
@@ -0,0 +1,221 @@
+#pragma once
+#include <iostream>
+#include <memory>
+#include <stdexcept>
+
+namespace linked_list
+{
+
+// This is a very helpful website: https://www.cs.northwestern.edu/~riesbeck/programming/c++/stl-iterator-define.html#TOC23
+template<class T> class LinkedListIter;
+
+template<class T, class Allocator = std::allocator<T>>
+class LinkedList
+{
+public:
+    friend class LinkedListIter<T>;
+    typedef LinkedListIter<T> iterator;
+
+    // Basic assignment constructors
+    LinkedList() = default;
+    LinkedList(const T& val) // TODO T&& val ?
+    {
+        std::cout << "Basic construction: from " << val << std::endl;
+        data = std::make_unique<T>(val);
+        mSize += 1;
+    }
+    // Copy construction
+    LinkedList(const LinkedList& src)
+    {
+        if (src.data)
+        {
+            std::cout << "Copy construction: from " << src.front() << std::endl;
+            data = std::make_unique<T>(src.front());
+            std::cout << "copy -- set data" << std::endl;
+        }
+        else
+        {
+            std::cout << "Copy construction: from empty" << std::endl;
+        }
+
+        if (src.next)
+        {
+            std::cout << "copy -- setting next" << std::endl;
+            next = std::make_unique<LinkedList<T>>(*src.next.get());
+            std::cout << "copy -- set next" << std::endl;
+        }
+        else
+            std::cout << "copy -- no next" << std::endl;
+        mSize = src.mSize;
+        std::cout << "copy -- set size" << std::endl;
+    }
+    // Move construction
+    LinkedList(LinkedList&& src)
+    {
+        std::cout << "Move construction: from " << src.front() << std::endl;
+        data = std::move(src.data);
+        next = std::move(src.next);
+        mSize = src.mSize;
+    }
+    // Copy assignment
+    LinkedList& operator=(const LinkedList& src)
+    {
+        std::cout << "Copy assignment: from " << src.front() << std::endl;
+        data = std::make_unique<T>(src.front());
+        if (src.next)
+            next = std::make_unique<LinkedList<T>>(*src.next.get());
+        mSize = src.mSize;
+        return *this;
+    }
+    // Move assignment
+    LinkedList& operator=(LinkedList&& src)
+    {
+        std::cout << "Move assignment: from " << src.front() << std::endl;
+        data = std::move(src.data);
+        next = std::move(src.next);
+        mSize = src.mSize;
+        return *this;
+    }
+
+    bool operator==(LinkedList& cmp) const
```

Comment:
I'm having trouble with `==` because the compiler insists on calling the binary version.